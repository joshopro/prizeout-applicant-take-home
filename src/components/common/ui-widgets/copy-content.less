.copy-container {
    display: block;
    margin-bottom: 1rem;

    &.copy-success {
        .copy-label-container {
            .copied {
                opacity: 1;
                visibility: visible;
            }
        }
        .copy-value-container {
            span {
                i {
                    background-color: #2BC86E;
                }
            }
        }
    }

    .copy-label-container {
        display: flex;
        font-size: .75rem;
        justify-content: space-between;

        .copy-label {
            color: #808997;
            margin: 0 0 .33rem 0!important;
        }

        .copied {
            color: #2BC86E;
            opacity: 0;
            transition: opacity 0.3s, visibility 0.3s;
            visibility: hidden;
        }
    }

    .copy-value-container {
        border: 2px solid #B7BBC0;
        border-radius: .5rem;
        cursor: pointer;
        display: flex;
        justify-content: space-between;
        padding: .33rem .33rem .33rem .5rem;
        align-items: center;
        align-self: center;

        .copy-value {
            color: #021431;
            font-size: 1rem;
        }

        span {
            i {
                color: #fff;
                background-color: #99A1AC;
                border-radius: 8px;
                padding: .5rem;
                transition: background-color 0.3s;
            }
        }
    }
}