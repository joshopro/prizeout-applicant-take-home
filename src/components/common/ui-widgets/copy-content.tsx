import React, { useState } from 'react';
import Classnames from 'classnames';
import PropTypes from 'prop-types';
import useIsMounted from 'react-is-mounted-hook';

import './copy-content.less';

interface CopyContentProps {
    content: string;
    title: string;
}

export const CopyContent: React.FC<CopyContentProps> = ({ content, title }): React.ReactElement => {
    const [contentCopied, setContentCopied] = useState(false),
        classes: string = Classnames('copy-container', { 'copy-success': contentCopied }),
        copyContent = () => {
            if (copiedTimer) {
                clearTimeout(copiedTimer);
            }
            const elem = document.createElement('textarea');
            elem.value = content;
            document.body.appendChild(elem);
            elem.select();
            document.execCommand('copy');
            document.body.removeChild(elem);
            setContentCopied(true);
            copiedTimer = setTimeout(() => {
                if (isMounted()) {
                    setContentCopied(false);
                }
            }, 2500);
        },
        isMounted = useIsMounted();

    let copiedTimer: ReturnType<typeof setTimeout>;

    return (
        <div className={classes}>
            <div className="copy-label-container">
                <p className="copy-label">{title}</p>
                <p className="copied">Copied!</p>
            </div>
            <div className="copy-value-container" onClick={copyContent}>
                <p className="copy-value">{content}</p>
                <span>
                    <i className="fas fa-copy" title="copy-icon"></i>
                </span>
            </div>
        </div>
    );
};

CopyContent.propTypes = {
    content: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
};
