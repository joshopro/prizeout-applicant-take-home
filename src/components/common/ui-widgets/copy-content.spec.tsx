import { render, screen } from '../../../testing/test-utils';
import userEvent from '@testing-library/user-event';
import React from 'react';
import '@testing-library/jest-dom';

import { CopyContent } from './copy-content';

jest.useFakeTimers();

describe('Test Copy Component', () => {
    const content = `Copy Me Content`;
    const title = `Copy Title`;
    const mockDocumentExecCommand = jest.fn((): boolean => true);

    beforeEach(() => {
        document.execCommand = mockDocumentExecCommand;
    });

    test('Component matches snapshot', () => {
        const { asFragment } = render(<CopyContent content={content} title={title} />);
        expect(asFragment()).toMatchSnapshot();
    });

    // test copied message shows
    test('Copied message shows when clicking on the container', () => {
        render(<CopyContent content={content} title={title} />);

        userEvent.click(screen.queryByTitle('copy-icon'), { button: 0 });
        expect(screen.getByText('Copied!')).toBeVisible();
    });

    test('Copy to clipboard function has been called', async () => {
        render(<CopyContent content={content} title={title} />);
        userEvent.click(screen.queryByTitle('copy-icon'), { button: 0 });
        expect(mockDocumentExecCommand.mock.calls.length).toBe(1);
    });

    // test('Copied message hides after timeout', async () => {
    //     act(() => {
    //         render(<CopyContent content={content} title={title} />);
    //     });

    //     userEvent.click(screen.queryByTitle('copy-icon'), { button: 0 });

    //     act(() => {
    //         jest.runAllTimers();
    //     });

    //     expect(screen.getByText('Copied!')).not.toBeVisible();
    // });
});
