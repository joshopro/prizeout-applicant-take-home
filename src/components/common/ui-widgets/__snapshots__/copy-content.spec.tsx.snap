// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Copy Component Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="copy-container"
  >
    <div
      class="copy-label-container"
    >
      <p
        class="copy-label"
      >
        Copy Title
      </p>
      <p
        class="copied"
      >
        Copied!
      </p>
    </div>
    <div
      class="copy-value-container"
    >
      <p
        class="copy-value"
      >
        Copy Me Content
      </p>
      <span>
        <i
          class="fas fa-copy"
          title="copy-icon"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
